### ISSUE TITLE
 [FEATURE] Harvester node driver to support multi-nics and multi-disks

### ISSUE THREAD

**Issue by naeemarsalan** [harvester/harvester#2426]
**Describe the bug**
When a cluster is provisioned with Rancher through Harvester node driver with multi nic. Only 1 nic is present on the VM.

Harvester Template
```apiVersion: harvesterhci.io/v1beta1
kind: VirtualMachineTemplateVersion
metadata:
  annotations:
    template-version.harvesterhci.io/customName: 7E0LP6IQ1U
  creationTimestamp: "2022-06-27T18:17:17Z"
  generateName: raw-image-base-template-
  generation: 2
  labels:
    template.harvesterhci.io/templateID: raw-image-base-template
  name: raw-image-base-template-pwdxh
  namespace: harvester-public
  ownerReferences:
  - apiVersion: harvesterhci.io/v1beta1
    blockOwnerDeletion: true
    controller: true
    kind: VirtualMachineTemplate
    name: raw-image-base-template
    uid: 6f09f440-0fe9-4526-bd89-6e4ab3f4b395
  resourceVersion: "5208595"
  uid: 52e1a42b-549d-47dc-b172-a1fcd1882c49
spec:
  templateId: harvester-public/raw-image-base-template
  vm:
    metadata:
      annotations:
        harvesterhci.io/volumeClaimTemplates: '[{"metadata":{"name":"pvc-rootdisk"},"spec":{"accessModes":["ReadWriteMany"],"resources":{"requests":{"storage":"10Gi"}},"volumeMode":"Block"}}]'
      creationTimestamp: null
      labels:
        harvesterhci.io/os: linux
    spec:
      runStrategy: RerunOnFailure
      template:
        metadata:
          annotations:
            harvesterhci.io/sshNames: '[]'
          creationTimestamp: null
          labels:
            harvesterhci.io/vmName: ""
        spec:
          domain:
            cpu:
              cores: 1
            devices:
              disks:
              - bootOrder: 1
                disk:
                  bus: virtio
                name: rootdisk
              interfaces:
              - masquerade: {}
                name: default
              - bridge: {}
                model: virtio
                name: nic-1
            features:
              acpi:
                enabled: true
            machine:
              type: ""
            resources:
              limits:
                cpu: "1"
                memory: 2Gi
          evictionStrategy: LiveMigrate
          networks:
          - name: default
            pod: {}
          - multus:
              networkName: default/vlan2-87
            name: nic-1
          volumes:
          - name: rootdisk
            persistentVolumeClaim:
              claimName: pvc-rootdisk
```

[cloud-init-output.log](https://github.com/harvester/harvester/files/8994775/cloud-init-output.log)
[cloud-init.log](https://github.com/harvester/harvester/files/8994776/cloud-init.log)

**To Reproduce**
1. Go to Harvester Create template with multi Nics
2. Go to Rancher use Harvester Node Driver use the following settings for
 
```
Network:
version: 1
config:
 - type: physical
  name: enp1s0 # name is varies upon OS image
  subnets:
   - type: dhcp
 - type: physical
  name: enp2s0
  subnets:
   - type: DHCP
 - type: physical
  name: enp3s0
  subnets:
   - type: DHCP 
 ```
3. Choose multus calico CNI provision cluster
4. VM is provisioned with 1 NIC.

**Expected behavior**
VM to multi Nics from template.

**Environment**
 - Harvester ISO version: 1.0.2
 - Underlying Infrastructure (e.g. Baremetal with Dell PowerEdge R630): Civo Baremetal

**Comment by FrankYang0529** [comment]
Hi @naeemarsalan, may I know how you used Rancher with the Harvester template to create RKE2? On the Rancher dashboard, I can't find the option.

<img width="1055" alt="Screenshot 2022-12-05 at 2 22 56 PM" src="https://user-images.githubusercontent.com/4927361/205563576-9a290e70-5869-47d0-b3b5-5225833c8a40.png">

In harvester, I can use the Harvester template which has multi-nic to provision a VM. My network data is
```
network:
  version: 1
  config:
  - type: physical
    name: enp1s0
    subnets:
      - type: dhcp
  - type: physical
    name: enp2s0
    subnets:
      - type: dhcp
  - type: physical
    name: enp3s0
    subnets:
      - type: dhcp
```
<img width="1275" alt="Screenshot 2022-12-05 at 2 28 00 PM" src="https://user-images.githubusercontent.com/4927361/205564252-934dfc4b-b878-4480-8632-33acecbd7a62.png">

**Comment by TachunLin** [comment]
### Test Plan

* Guest cluster support: RKE1, RKE2 
* Rancher version: v2.7.0
* Harvester version: latest master  

#### Major check point:
1. Provision guest cluster (RKE1/RKE2)
2. Edit guest cluster, can support add multiple nics and multiple volumes 

#### Comparability Matrix 
| UI     | node driver     | Config     |
|--------|-----------------|------------|
| new UI | new node driver | new config |
| new UI | new node driver | old config |
| old UI | new node driver | old config |

#### Verification Steps
1. Use existing UI and Node driver to provision RKE1 and RKE2 guest cluster 
2. Update node driver to v0.6.0
3. Edit guest cluster, check the multi-nics and multi-disks
4. Update RKE1/RKE2 UI (through ui-dashboard, ui-index)
5. Change `ui-offline-preferred` to `Remote`
6. Edit guest cluster, check the multi-nics and multi-disks
7. Create another new RKE1 and RKE2 guest cluster 
8. Edit guest cluster, check the multi-nics and multi-disks

**Comment by TachunLin** [comment]
### RKE1 Test Result: PASS

| UI     | node driver     | Config     | Result |
|--------|-----------------|------------|--------|
| new UI | new node driver | new config | PASS   |
| new UI | new node driver | old config | PASS  |
| old UI | new node driver | old config | PASS  |

1. Use existing UI and Node driver to provision RKE1 and RKE2 guest cluster (K8s version: v1.24.8)
![image](https://user-images.githubusercontent.com/29251855/212816665-929cdf22-eec7-444e-9f07-43ecc0793aea.png)

2. Update node driver to v0.6.0
![image](https://user-images.githubusercontent.com/29251855/212817238-e150e50d-181e-4add-822c-98ca97201e3d.png)

![image](https://user-images.githubusercontent.com/29251855/212817326-9501cf70-698a-4101-8177-199d03786689.png)


3. Edit RKE1 node template with new value 
4. Create a new RKE1 cluster with updated node driver v0.6.0 on updated RKE1 template 
5. Check can provision RKE1 cluster correctly

6. Update RKE1/RKE2 UI (through ui-dashboard, ui-index)
7. Change `ui-offline-preferred` to `Remote`
![image](https://user-images.githubusercontent.com/29251855/212831451-ce4510cc-2471-4dfd-8d90-e2cc03fcf269.png)
8. Edit existing RKE1 node template
9. Add new volume
![image](https://user-images.githubusercontent.com/29251855/212832032-99ad2f79-22c6-414c-a606-25036edae3b5.png)

10. Add new network
![image](https://user-images.githubusercontent.com/29251855/212832094-d98907a3-04f4-47b2-a0c3-391ca1b892f0.png)

11. Add network data for multiple-nics
  ```
  network:
    version: 1
    config:
     - type: physical
       name: enp1s0
       subnets:
       - type: dhcp
     - type: physical
       name: enp2s0
       subnets:
       - type: dhcp
  ```

![image](https://user-images.githubusercontent.com/29251855/212832232-946b4934-c960-4720-a5ef-16cf279821f7.png)


12. Check RKE1 guest cluster network and volume config
![image](https://user-images.githubusercontent.com/29251855/212834068-71997719-6ff5-430e-aa73-c00197127b70.png)

![image](https://user-images.githubusercontent.com/29251855/212834113-9b226272-775f-49a9-b774-a80293aa3e11.png)

![image](https://user-images.githubusercontent.com/29251855/212834144-9e879f44-2531-4295-9fce-f8db3161acb9.png)

13. Check can provision RKE1 guest cluster correctly with multi-nics and multi-volumes
![image](https://user-images.githubusercontent.com/29251855/212834782-51ac785d-3146-41ee-b1ce-63e93bf013c7.png)

14. Create a new RKE1 node template, adding multi-nics and multi-volumes
15. Create another new RKE1 guest cluster 
16. Check RKE1 guest cluster network and volume config
![image](https://user-images.githubusercontent.com/29251855/212837516-c0e08190-8e31-4f31-945a-329f49ac164a.png)


17. Check can provision RKE1 guest cluster correctly with multi-nics and multi-volumes
![image](https://user-images.githubusercontent.com/29251855/212837603-639afb76-e600-4370-b98f-e4c3139a5301.png)

18. Check can delete the RKE1 guest cluster with new node driver

**Comment by TachunLin** [comment]
### RKE2 Test Result: FAILED

| UI     | node driver     | Config     | Result |
|--------|-----------------|------------|--------|
| new UI | new node driver | new config | FAILED    |
| new UI | new node driver | old config | FAILED  |
| old UI | new node driver | old config |   |

We found when updating existing config of the RKE2 guest cluster in the following cases:
1. New node driver with new UI config
2. New node driver with old UI config 




The previous guest VM will stuck in `deleting` state, with error
  ```
   Error removing host "rke2-olc-pool1-9b6c33f8-8kkjt": json: cannot unmarshal string into Go struct field Driver.DiskSize of type int
  ```
![image](https://user-images.githubusercontent.com/29251855/212871050-f9227fd5-e72e-4161-918d-ea812bfea148.png)

![image](https://user-images.githubusercontent.com/29251855/212870703-5cb827ff-7545-4280-9e58-4a6ac27d30b5.png)
#### Verification Steps
1. Use existing UI and Node driver to provision RKE2 guest cluster (K8s version: v1.24.8)
4. Update node driver to v0.6.0
5. Update RKE1/RKE2 UI (through ui-dashboard, ui-index)
6. Change `ui-offline-preferred` to `Remote`
![image](https://user-images.githubusercontent.com/29251855/212831451-ce4510cc-2471-4dfd-8d90-e2cc03fcf269.png)
7. Edit existing RKE2 cluster config 
8. Add new volume
![image](https://user-images.githubusercontent.com/29251855/212865962-f7f4e61d-6e86-41c2-808c-75edc81f7a68.png)


10. Add new network
![image](https://user-images.githubusercontent.com/29251855/212866011-84b5c1b8-f403-4f57-9aaa-d978700cdecd.png)


11. Add network data for multiple-nics
  ```
  network:
    version: 1
    config:
     - type: physical
       name: enp1s0
       subnets:
       - type: dhcp
     - type: physical
       name: enp2s0
       subnets:
       - type: dhcp
  ```

![image](https://user-images.githubusercontent.com/29251855/212866167-e772a7c6-89aa-41e8-8a8c-dc05914b5720.png)


12. Check RKE2 guest cluster network and volume config after updating finish
![image](https://user-images.githubusercontent.com/29251855/212869267-f08cf1b6-faba-4f91-8b2d-6767e0536a7c.png)


13. But `failed` to update RKE2 guest cluster correctly with multi-nics and multi-volumes,
![image](https://user-images.githubusercontent.com/29251855/212869370-0c7425be-c4d3-40dd-9cd7-7e330d890c3d.png)

14. Unable to delete the previous VM

**Comment by TachunLin** [comment]
Re-verified with updated node driver https://github.com/futuretea/docker-machine-driver-harvester/releases/download/v0.0.0-f8f2ef2/docker-machine-driver-harvester-amd64.tar.gz 

After we updated RKE2 cluster config, adding multi-nics and multi-disks. 
The previous guest cluster vm can be correctly removed after 15 minutes around. 

| UI     | node driver     | Config     | Result |
|--------|-----------------|------------|--------|
| new UI | new node driver | new config |     |
| new UI | new node driver | old config | PASS  |
| old UI | new node driver | old config |   |

![image](https://user-images.githubusercontent.com/29251855/212887713-58837e83-d883-4057-9b7c-198ba97ec909.png)

**Comment by TachunLin** [comment]
Move to implement and plan to verify after related PR merged.

**Comment by futuretea** [comment]
@TachunLin released v0.6.1

Download URL: https://releases.rancher.com/harvester-node-driver/v0.6.1/docker-machine-driver-harvester-amd64.tar.gz
checksum: 0c86c5797f6ef13881632a4077a70f62cb00937bbe8896c7fda72bc50abb0cea

**Comment by TachunLin** [comment]
### RKE2 Test Result: PASS 

Verified with node driver v0.6.1 can pass the following test scenarios:

1. Can update existing RKE2 guest cluster using new node driver
2. Can add multi-nics and multi-disks of old RKE2 guest cluster using new node driver and ui config 
3. Can create new RKE2 guest cluster with multi-nics and multi-disks correctly
4. Can delete RKE2 guest cluster correctly

| UI     | node driver     | Config     | Result |
|--------|-----------------|------------|--------|
| new UI | new node driver | new config | PASS |
| new UI | new node driver | old config | PASS |
| old UI | new node driver | old config | PASS  |


#### Verification Steps
1. Use existing UI and Node driver to provision RKE2 guest cluster (K8s version: v1.24.8)
5. Update node driver to v0.6.1
6. Edit guest cluster config, change cpu and memory settings
7. Check can update the guest cluster with new node driver
![image](https://user-images.githubusercontent.com/29251855/213098846-62c5f4d3-fe45-4de7-ba39-b0d28152a0f5.png)

8. Check can correctly delete the previous guest cluster
![image](https://user-images.githubusercontent.com/29251855/213098108-57f552b1-ae85-4534-b68b-1ec375d27707.png)

![image](https://user-images.githubusercontent.com/29251855/213098891-2f0407ff-ba4e-4fc1-a40e-cb96ea40b764.png)


9. Update RKE1/RKE2 UI (through ui-dashboard, ui-index)
10. Change `ui-offline-preferred` to `Remote`
11. Edit existing RKE2 cluster config 
12. Add new volume
13. Add new network
14. Add network data for multiple-nics
  ```
  network:
    version: 1
    config:
     - type: physical
       name: enp1s0
       subnets:
       - type: dhcp
     - type: physical
       name: enp2s0
       subnets:
       - type: dhcp
  ```
15. Check RKE2 guest cluster network and volume config after updating finish
![image](https://user-images.githubusercontent.com/29251855/213103740-652d9228-a4bf-450c-bc76-1b9311bce1c4.png)

16. Check can update RKE2 guest cluster correctly with multi-nics and multi-volumes,
![image](https://user-images.githubusercontent.com/29251855/213103800-e5955ce6-716a-46a8-9748-ada606ccf509.png)
![image](https://user-images.githubusercontent.com/29251855/213103854-63b13cc4-ac30-4c84-8d05-69a017a72d59.png)


17. Check can delete the previous guest cluster VM
![image](https://user-images.githubusercontent.com/29251855/213103697-64fe0c9c-346e-4a27-9ffe-ffcfc8d025a3.png)

18. Check can delete the RKE2 cluster with new node driver
19. Create a new RKE2 guest cluster with new node driver and new UI config 
20. Add new volume
21. Add new network
22. Add network data for multiple-nics
23. Check can create new RKE2 guest cluster correctly with multi-nics and multi-volumes,
![image](https://user-images.githubusercontent.com/29251855/213105827-7075e1d8-208f-431a-81d8-68a7ce082b58.png)
![image](https://user-images.githubusercontent.com/29251855/213105855-5e8054ce-37f3-40a7-abe5-f9e01ed60b01.png)
![image](https://user-images.githubusercontent.com/29251855/213105933-665f0197-8f30-4e55-9e62-1ad9a505630c.png)
![image](https://user-images.githubusercontent.com/29251855/213105967-85db350d-ec5a-440d-8aac-8c6184660713.png)

**Comment by TachunLin** [comment]
Verified **fixed** on master-14bbee2c-head (1/18). Close this issue.

### Result
Verified with node driver v0.6.1 can pass the following test scenarios:

1. Can update existing RKE2 guest cluster using new node driver
2. Can add multi-nics and multi-disks of old RKE2 guest cluster using new node driver and ui config 
3. Can create new RKE2 guest cluster with multi-nics and multi-disks correctly
4. Can delete RKE2 guest cluster correctly

Verified with node driver v0.6.0 can pass the following test scenarios:
1. With new node driver only, can update existing RKE1 node template to create RKE1 cluster correctly
2. With new node driver and ui config, can add multi-nics and multi-disks to node template and create RKE1 culster correctly
3. With new node driver and ui config, can create new node template with multi-nics and multi-disks and create RKE1 cluster correctly

### Test Information
* Test Environment: 1 node harvester on local kvm machine
* Harvester version: `master-14bbee2c-head` 
* Rancher version: v2.7.0
* Node driver version: v0.6.1
* ui-index: https://harvester-dev.oss-cn-hangzhou.aliyuncs.com/ui/latest2/index.html
* ui-dashboard-index: https://harvester-dev.oss-cn-hangzhou.aliyuncs.com/dashboard/support-multi-nic/index.html
* ui-offline-preferred: Remote

### Verify Steps
RKE1: https://github.com/harvester/harvester/issues/2426#issuecomment-1384951843
RKE2: https://github.com/harvester/harvester/issues/2426#issuecomment-1386572469

**Comment by LucasSaintarbor** [comment]
Hi @futuretea @WuJun2016 @TachunLin, just following up on any issue with a `require/doc` label for the 1.2.0 release. Does this issue require any of the following? 

- [ ] New documentation/KB
- [ ] Updated documentation/KB
- [ ] Release note
- [x] None of the above

If it does, may you please open a PR or share details on what needs to be created or updated?
Thanks!
